torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,196 @@
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length: int = total_length
+        self.remaining_length: int = total_length
+        self._seed = seed
+        self.keys: List[T] = list(weights.keys())
+        self.key_to_index: Dict[T, int] = {k: i for i, k in enumerate(self.keys)}
+        self.norm_weights: List[float] = self.normalize_weights([weights[k] for k in self.keys], total_length)
+        self.weights: List[float] = self.norm_weights.copy()
+        self._rng = random.Random(self._seed)
+        self._lengths: List[int] = []
+
+    def draw(self) -> T:
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        index = self.key_to_index[selected_key]
+        self.weights[index] -= 1
+        self.remaining_length -= 1
+        if self.weights[index] < 0:
+            self.weights[index] = 0
+            self.weights = self.normalize_weights(self.weights, self.remaining_length)
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights: List[float], total_length: int) -> List[float]:
+        """
+        Given a ``List`` of weights, normalize them according to ``total_length``.
+        """
+        total_weight = sum(weights)
+        return [float(w) * total_length / total_weight for w in weights]
+
+    def reset(self) -> None:
+        self._rng = random.Random(self._seed)
+        self.weights = self.norm_weights.copy()
+        self.remaining_length = self.total_length
+
+    def override_seed(self, seed):
+        """
+        Update the `seed`. The new `seed` will be used in the next iteration.
+        """
+        self._seed = seed
+        return self
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            self._rng.getstate(),
+        )
+        return state
+
+    def __setstate__(self, state):
+        (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            rng_state,
+        ) = state
+        self._rng = random.Random()
+        self._rng.setstate(rng_state)
+
+    def get_length(self, target: T) -> int:
+        if not self._lengths:
+            if all(w.is_integer() for w in self.norm_weights) and sum(self.norm_weights) == self.total_length:
```

Comment:
We can potentially assert this during `__init__`, such that if the normalized weights aren't integer or cannot sum up to `total_length`, we will raise exception.